www/js/services/indicoApiService.js

```diff
@@ -0,0 +1,181 @@
+        const {access_token: accessToken} = getSearchParams(authBrowserResult, '#');
+
+        resolve({...server, token: accessToken});
+      } catch (e) {
+        reject(e);
+      }
+    });
+  };
+
+  const authPkceFlow = server => {
+    return new Promise(async (resolve, reject) => {
+      const state = generateRandomString(128);
+      const codeVerifier = generateRandomString(128);
+      const codeChallenge = await generateCodeChallenge(codeVerifier);
+      const authUrl = buildUrl(server.authUrl, {
+        response_type: 'code',
+        client_id: server.clientId,
+        redirect_uri: server.callbackUrl,
+        state,
+        scope: server.scope,
+        code_challenge: codeChallenge,
+        code_challenge_method: 'S256',
+      });
+
+      try {
+        const authBrowserResult = await browserSignOn(authUrl.href, server.callbackUrl);
+        const {code, state: responseState, error} = getSearchParams(authBrowserResult);
+
+        if (error || !code || state !== responseState) reject();
+
+        const tokenResponse = await fetch(server.tokenUrl, {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
+          },
+          body: urlEncodeForm({
+            grant_type: 'authorization_code',
+            client_id: server.clientId,
+            redirect_uri: server.callbackUrl,
+            code,
+            code_verifier: codeVerifier,
+          }),
+        });
+        const data = await tokenResponse.json();
+
+        resolve({...server, token: data.access_token});
+      } catch (e) {
+        reject(e);
+      }
+    });
+  };
+
+  function authenticate(server, usePkce = false) {
+    if (usePkce) {
+      return authPkceFlow(server);
+    } else {
+      return authImplicitFlow(server);
+    }
+  }
+
+  function getServerData(server, version) {
+    const serverDataVersion = {
+      1: server => ({
+        serverId: getKey(server.base_url || server.baseUrl),
+        clientId: server.consumer_key || server.consumerKey,
+        baseUrl: server.base_url || server.baseUrl,
+        authUrl: server.auth_url,
+        tokenUrl: server.token_url,
+        scope: 'registrants',
+        callbackUrl: 'http://localhost',
+      }),
+      2: async server => {
+        const oAuthDiscoveryUrl = `${server.base_url}/.well-known/oauth-authorization-server`;
+        const res = await fetch(oAuthDiscoveryUrl);
+        const data = await res.json();
+
+        return {
+          serverId: getKey(server.base_url),
+          clientId: server.client_id,
+          baseUrl: server.base_url,
+          authUrl: data.authorization_endpoint,
+          tokenUrl: data.token_endpoint,
+          scope: server.scope,
+          usePkce: true,
+          callbackUrl: 'http://localhost',
+        };
+      },
+      0: () => {
+        throw new Error('Incompatible QR.');
+      },
```

Comment:
This won't handle a higher version number coming from Indico. switch or if/elseif may be more appropriate here